---
title: "Bioinformatics methods"
output: word_document
bibliography: refs.bib
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, cache=TRUE, message = F)
suppressMessages(library(DESeq2))
suppressMessages(library(limma))
suppressMessages(library(data.table))
suppressMessages(library(Heatplus))
```

## Read mapping

We used Ensembl version 89 to obtain the mouse genome in FASTA format (GRCm38) and gene models in gff. The bowtie2 indices were built from the genome index as

```bash
bowtie2-build GRCm38.fa GRCm38
tophat2 -G GRCm38.genes.gff --transcriptome-index=GRCm38.genes GRCm38
```

Raw RNAseq paired-end reads were aligned to the genome using default options in Bowtie2-Tophat2&nbsp;[@Kim2013] pipeline.
 
```bash
tophat2  --no-novel-juncs -T -p 8 -o ../data/processed/alignments/sample  --transcriptome-index ../data/genome/genes.GRCm38.89    ../data/genome/GRCm38 ../data/raw-data/sample_R1.fastq.gz ../data/raw/sample_R2.fastq.gz
```

The resulting bam file for accepted hits generated from by tophat2 for each sample was then used to count reads mapping to individual genes using HTseq.

```bash
htseq-count -s reverse -f bam sample/accepted_hits.bam GRCM38.genes.gff > counts/sample
```



## Differential Expression

Differential expression was estimated using DESeq2.

### Knockout model

Differential expression between the wild type and e-cadherin knockout samples was calculated by reading in the counts generated by HTSeq&nbsp;[@Anders2015].

```{r}
samplesTable <- data.frame(samplenames=c("KO1","KO2","KO3","KO4","KO5","WT1","WT2","WT3","WT5"),
                  samplefiles=c("VPad-KO1_S5_","VPad-KO2_S6_","VPad-KO3_S7_","VPad-KO4_S8_","VPad-KO5_S9_","VPad-WT1_S1_","VPad-WT2_S2_","VPad-WT3_S3_","VPad-WT5_S4_"),
                  genotype=c("KO","KO","KO","KO","KO","WT","WT","WT","WT"))

samplesTable$genotype <- relevel(x = samplesTable$genotype, ref = "WT")
```

```{r, echo=TRUE}
dds <- DESeqDataSetFromHTSeqCount(samplesTable, directory = "htseq_output/", design = ~ genotype)
```
where the samples are annotated with the particular genotype (knock-out or wild-type) as
```{r}
knitr::kable(samplesTable)
```

DESeq2 estimates the sample sizes (sequencing depth per sample), dispersion (variance in read counts for genes for the negative binomial model), automatically to compute the differential expression fold changes and p-values.

```{r, echo=TRUE}
dds <- DESeq(dds)
results.KO.DESeq2 <- results(dds, contrast = c("genotype", "KO", "WT"))
```


```{r}
results.KO.DESeq2$Min.reads = rowMin(assay(dds))
results.KO.DESeq2$KO.reads = rowSums(assay(dds)[,c("KO1","KO2","KO3","KO4","KO5")])
results.KO.DESeq2$WT.reads = rowSums(assay(dds)[,c("WT1","WT2","WT3","WT5")])
```

### E-cadherin linear expression model

We also computed the differential expression of each gene according to its expression as it relates to the Cdh1 expression in each sample. First, the E-cadherin expression reads per million were calculated, and this was used
as the design matrix in to calculate the differential expression.

```{r, echo=TRUE}
samplesTable$ecad_rpm = counts(dds)["gene:ENSMUSG00000000303",]/colSums(counts(dds)) * 1e6
```


### E-cadherin linear expression normalized

We scale the Cdh1 reads per million to annotate each sample so that the WT samples have a mean 0 and KO samples have mean 1 Cdh1 value. Note that this is effectively inverting the Cdh1 expression pattern, but this variable gives us fold changes in the same scale and direction as the KO against WT comparison. This gives us:

```{r}
samplesTable <- as.data.table(samplesTable)
samplesTable$norm_ecad = samplesTable$ecad_rpm - mean(samplesTable[grep(pattern = "WT", samplenames), ecad_rpm])
samplesTable$norm_ecad = samplesTable$norm_ecad / mean(samplesTable[grep(pattern = "KO", samplenames), norm_ecad])
knitr::kable(samplesTable)
```

We compute the linear model DE using this normalized ecadherin expression:

```{r, echo=TRUE}
dds_norm_ecad <- DESeqDataSetFromHTSeqCount(samplesTable, directory = "htseq_output/", design = ~ norm_ecad)

dds_norm_ecad <- DESeq(dds_norm_ecad)

results.norm_ecad.DESeq2 <- results(dds_norm_ecad)
```


```{r name-mapping}
suppressMessages(library(ggplot2))
suppressMessages(library(biomaRt))

ensembl <-  useMart("ensembl", dataset = "mmusculus_gene_ensembl")
e2n <- as.data.table(getBM(attributes = c("ensembl_gene_id", "mgi_symbol", "description"), mart = ensembl))
e2n <- unique(e2n)
```


```{r put-all-results-together}

allresults <- data.frame(geneid = rownames(results.KO.DESeq2), 
                         pvalue.KO.DESeq2 = results.KO.DESeq2$pvalue,
                         pvalue.ecad_norm.DESeq2 = results.norm_ecad.DESeq2$pvalue,
                         l2fc.KO.DESeq2 = results.KO.DESeq2$log2FoldChange,
                         l2fc.ecad_norm.DESeq2 = results.norm_ecad.DESeq2$log2FoldChange,
                         total.KO.reads = rowSums(counts(dds)[, grep(pattern = "KO", colnames(counts(dds)))]),
                         total.WT.reads = rowSums(counts(dds)[, grep(pattern = "WT", colnames(counts(dds)))]),
                         stringsAsFactors = F
                         )


allresults <- as.data.table(allresults)


```





```{r}
library(openxlsx)
allresults$geneid <- sapply(X = strsplit(allresults$geneid, split = ":"), FUN = function(x) x[2])
setkey(allresults, geneid)
setkey(e2n, ensembl_gene_id)
allresults <- merge(allresults, e2n, by.x = "geneid", by.y = "ensembl_gene_id", all.x=T)
setkey(allresults, geneid)

```


## Multiple testing correction and filtering un-mapped and low read count genes

DESeq2 does not compute p-values for genes with no reads, and for the KO-WT comparison, also for genes with outlying read-counts in one of the samples. Outlying read counts in one of the samples are defined by the default parameter in DESeq2, corresponding to a cutoff of one sample's read count being farther than the 0.99 quantile. For more details, see @Love2014 and the associated manual.

We manually set the p-values to 1 for all the genes for which DESeq2 did not calculate p-values. This avoids missing values in the tables, and because it is not possible to conclusively reject the null hypothesis.

```{r get-genes-to-filter}
badGenesPVal = allresults[is.na(pvalue.KO.DESeq2), geneid]
badGenesFDR = union(badGenesPVal, allresults[(total.KO.reads + total.WT.reads) < 10, geneid])
goodGenesFDR = setdiff(allresults$geneid, badGenesFDR)


```


Out of the total `r dim(dds)[1]` genes, `r length(badGenesPVal)` did not have a p-value in the KO-WT comparison, and `r sum(! is.na(allresults$pvalue.KO.DESeq2))` had a valid p-value. For the Cdh1 linear model, `r sum(is.na(allresults$pvalue.ecad_norm.DESeq2))` had missing p-values and `r sum(! is.na(allresults$pvalue.ecad_norm.DESeq2))` genes had a valid p-value.

For multiple testing correction, we calculated the false discovery rate for all genes not having outlier values and with total read counts greater than or equal to 10. This limited the effective number of tests (genes) used for FDR calculations to `r length(goodGenesFDR)`.

```{r calculate-fdr}

allresults[is.na(pvalue.KO.DESeq2), pvalue.KO.DESeq2 := 1]
allresults[is.na(pvalue.ecad_norm.DESeq2), pvalue.ecad_norm.DESeq2 := 1]


allresults[goodGenesFDR, FDR.KO := p.adjust(pvalue.KO.DESeq2, method = "fdr")]
allresults <- allresults[goodGenesFDR, FDR.ecad_norm := p.adjust(pvalue.ecad_norm.DESeq2, method = "fdr")]


allresults[badGenesFDR, FDR.KO := 1]
allresults <- allresults[badGenesFDR, FDR.ecad_norm := 1]
```




# Gene set enrichment

We downloaded the Msigdb gene sets to test for enrichment of differentially expressed genes. The mouse ensembl gene ids were converted to Human HGNC symbols to map the genes to Msigdb gene sets. (hat tip: Ewald lab).

```{r human-mouse-conversion}
library(biomaRt)
human = useMart("ensembl", dataset = "hsapiens_gene_ensembl")
mouse = useMart("ensembl", dataset = "mmusculus_gene_ensembl")
mousegenes <- allresults$geneid
hggenes <- getLDS(attributes = c("ensembl_gene_id"),
                  filters = "ensembl_gene_id", values = mousegenes,
                  mart = mouse, attributesL = c("hgnc_symbol"), 
                  martL = human, uniqueRows=T)

hggenes <- as.data.table(hggenes)
setkey(hggenes, HGNC.symbol)
```

Z-scores were calculated for each gene based on the p-value of differentially expression. z-scores were defined as abs(qnorm(pval/2)) * sign(log2FoldChange). For all genes with zero read counts or those that DESeq2 filtered out and a p-value of fold change estimated was not calculated, the z-score is zero.

```{r calculate-z-score}
allresults[, z.KO := abs(qnorm(pvalue.KO.DESeq2/2)) * sign(l2fc.KO.DESeq2)]
allresults[, z.ecad_norm := abs(qnorm(pvalue.ecad_norm.DESeq2/2)) * sign(l2fc.ecad_norm.DESeq2)]

allresults <- allresults[is.na(z.KO), z.KO := 0]
allresults <- allresults[is.na(z.ecad_norm), z.ecad_norm := 0]
```

```{r export-excel}
allresults <- allresults[order(allresults$pvalue.ecad_norm.DESeq2, decreasing = F),]
openxlsx::write.xlsx(allresults, file = "allresults.xlsx", 
                     sheetName = "DE Gene-level results", overwrite=TRUE,
                     asTable = T)


allresults <- allresults[order(allresults$pvalue.ecad_norm.DESeq2, decreasing = F),]
openxlsx::write.xlsx(allresults[, .(geneid, mgi_symbol, pvalue.ecad_norm.DESeq2, l2fc.ecad_norm.DESeq2) ],
                     file = "GeneResultsForVeena.xlsx",
                     sheetName = "DE Gene-level results", overwrite=TRUE,
                     asTable = T)


```


```{r add-human-symbols}
setkey(allresults, geneid)
setkey(hggenes, Gene.stable.ID)
allresults <- merge(x = allresults, y = hggenes, by.x="geneid", by.y="Gene.stable.ID", all.x=TRUE)
```


```{r write-z-scores-KO}
write.table(x = allresults[! is.na(HGNC.symbol), .(HGNC.symbol, z.KO)],
            file = "set_results/KO.zscores.txt",
            row.names = F,
            col.names = c("gene", "zscore"),
            quote = F,
            sep = "\t")
```

```{r write-zscores-norm-ecad}
write.table(x = allresults[! is.na(HGNC.symbol), .(HGNC.symbol, z.ecad_norm)],
            file = "set_results/norm_ecad.zscores.txt",
            row.names = F,
            col.names = c("gene", "zscore"),
            quote = F,
            sep = "\t")
```

The enrichment of DE genes in a gene set was calculated as the p-value of the t-test comparing the absolute value of the z-score between the in-set and out of set genes. Therefore, a gene set with genes that are highly regulated, but in both directions, will still be called significant.


```{r define-gsea}
run_gsea <- function(zscorefile, lsetfile="c2",  outputfile, dataprefix="set_results/", setprefix="set_results/", setsuffix=".all.v6.0.symbols.gmt")
{
  pypath <- function()
  {
    for(candidate in c("C:\\ProgramData\\Anaconda3\\python.exe", "python", "python.exe")) {
      if (system(command = candidate, ignore.stderr = T, ignore.stdout = T)==0) {
        return(candidate)
      }
    }
  }
  if (! missing(outputfile)) outfile=paste0(dataprefix, outputfile)
  else outfile=paste0(dataprefix, zscorefile, "_",lsetfile,"_enriched.txt")
  
  comm = paste0(pypath()," do_gsea.py ", dataprefix, zscorefile, " ", setprefix,lsetfile,setsuffix ," ",
                outfile)
  system(command = comm)
}
```

The gene set enrichment was carried out for both the KO-WT and Cdh1 linear analysis.


<!--
```{ do-gsea-KO}

run_gsea(zscorefile = "KO.zscores.txt", lsetfile = "c2", outputfile = "KO_c2.txt",
         dataprefix = "set_results/")
run_gsea(zscorefile = "KO.zscores.txt", lsetfile = "c5", outputfile = "KO_c5.txt",
         dataprefix = "set_results/")
run_gsea(zscorefile = "KO.zscores.txt", lsetfile = "h", outputfile = "KO_h.txt",
         dataprefix = "set_results/")
run_gsea(zscorefile = "KO.zscores.txt", lsetfile = "c6", outputfile = "KO_c6.txt",
         dataprefix = "set_results/")
```

```{ do-gsea-norm_ecad}

run_gsea(zscorefile = "norm_ecad.zscores.txt", lsetfile = "c2", outputfile = "norm_ecad_c2.txt",
         dataprefix = "set_results/")
run_gsea(zscorefile = "norm_ecad.zscores.txt", lsetfile = "c5", outputfile = "norm_ecad_c5.txt",
         dataprefix = "set_results/")
run_gsea(zscorefile = "norm_ecad.zscores.txt", lsetfile = "h", outputfile = "norm_ecad_h.txt",
         dataprefix = "set_results/")
run_gsea(zscorefile = "norm_ecad.zscores.txt", lsetfile = "c6", outputfile = "norm_ecad_c6.txt",
         dataprefix = "set_results/")
```

-->


```{r write-all-gsea}
library(openxlsx)
gsea_stats <- data.frame(DEMethod=NA, Design=NA, Collection=NA, "Number of sets"=NA, "Number pval < 0.05"=NA, "Number FDR < 0.05"=NA, check.names = F)

read_gsea <- function(setname, setdesc) {
  read_gsea_file <- function(filename) {
    set_results <- read.table(filename, sep = "\t", header = T, stringsAsFactors = F)
    set_results$padj <- p.adjust(set_results$pvalue, method = "fdr")
    set_results <- as.data.table(set_results)
    setkey(set_results, geneset)
    set_results
  }
  
  addsuffix <- function(df, suffix, exclude) {
    cn <- colnames(df)
    cn2 <- sapply(X = cn, FUN = function(cname) {
      if (cname %in% exclude) return(cname)
      paste0(cname, suffix)
    })
    colnames(df) <- cn2
    df
  }
  
  KO_set <- read_gsea_file(sprintf("set_results/KO_%s.txt",setname))
  gsea_stats <<- rbind(gsea_stats, list("DESeq2", "KO",paste0(setdesc, " (",setname,")"), dim(KO_set)[1], sum(KO_set$pvalue < 0.05),  sum(KO_set$padj < 0.05)))
  KO_set <- addsuffix(KO_set, ".KO.deseq2", c("geneset","comment","nset"))
  

  ecad_norm_set <- read_gsea_file(sprintf("set_results/norm_ecad_%s.txt",setname))
  gsea_stats <<- rbind(gsea_stats, list("DESeq2", "Cdh1 norm",paste0(setdesc, " (",setname,")"), dim(ecad_norm_set)[1], sum(ecad_norm_set$pvalue < 0.05),  sum(ecad_norm_set$padj < 0.05)))
  ecad_norm_set <- addsuffix(ecad_norm_set, ".Cdh1_norm.deseq2", c("geneset","comment","nset"))
  


  g_set_all <- merge(KO_set, ecad_norm_set, by=c("geneset","comment","nset"))

  return(g_set_all)
}



wb <- createWorkbook(creator = "Yasir Suhail")

collate_gsea_set <- function(setnameshort, setdescription, workbook) {
  g_set <- read_gsea(setname = setnameshort, setdesc = setdescription)
  shname = paste0(setnameshort," (",setdescription,")")
  addWorksheet(wb = workbook, sheetName = shname)
  # writeDataTable(wb = workbook, sheet = shname, x = g_set, colNames = T, rowNames = F, bandedRows = T, tableStyle = "TableStyleMedium2")
  writeDataTable(wb = workbook, sheet = shname, x = g_set[order(g_set$pvalue.Cdh1_norm.deseq2),], colNames = T, rowNames = F, bandedRows = T, tableStyle = "TableStyleMedium2")
  setColWidths(wb = workbook, sheet = shname, cols = c(1:ncol(g_set)), widths = 30, hidden = (! grepl(pattern = "(geneset)|(comment)|(padj)|(pval)|(nset)",x = colnames(g_set))))
}

addWorksheet(wb = wb, sheetName = "GSEA summary")

collate_gsea_set("c2","Curated Gene Sets",wb)
collate_gsea_set("c5","GO Gene Sets",wb)
collate_gsea_set("c6","Oncogenic Signatures",wb)
collate_gsea_set("h","Hallmarks",wb)


gsea_stats <- gsea_stats[-1,]
writeDataTable(wb = wb, sheet = "GSEA summary", x = gsea_stats, rowNames = F, colNames = T, bandedRows = T, tableStyle = "TableStyleMedium2")

saveWorkbook(wb = wb, file = "set_results/All KO vs Ecad GSEA.xlsx",overwrite = T)

```


# Visualization of the differential expression

The heatmap of the differentially expressed genes was plotted using the z-scores for the gene expression. The log-transformed read counts were transformed to z-scores using the following method. First, the read counts for each sample were divided by the total read count for the sample to normalize the sequencing depth/size of each sample. Then, these normalized counts were log transformed (accounting for low counts as log(1+$x$)). Subsequently, for each gene, the z-score was computed by subtracting the mean and dividing by the standard deviation of the log transformed data.

All the genes which were found to be significantly differentially expressed in either the KO-WT and Cdh1 linear model, where significance is defined as the (un-corrected) p-value of 0.05 or less, were included in the heatmap.


To illustrate the expression of genes involved in the epithelial-mesenchymal transition (EMT), only a select subset of genes were included in the heatmap. The normalization procedure to calculate z-scores included the total set of genes, before selecting the differentially expressed, or EMT subset.


```{r heatmatrix, echo=TRUE}
readsMatrix <- assay(dds)
readsSizeCorrected <- readsMatrix / colSums(readsMatrix)

readsLogMatrix <- log(1 + readsSizeCorrected)

GeneMeans <- rowMeans(readsLogMatrix)
GeneSDs <- rowSds(readsLogMatrix)

zMatrix <- (readsLogMatrix - GeneMeans)/GeneSDs
```


# Results


## Differential expression

```{r heatmap-pval.05, include=FALSE, message=FALSE}
annosets = c("KEGG_WNT_SIGNALING_PATHWAY","KEGG_APOPTOSIS","KEGG_TGF_BETA_SIGNALING_PATHWAY")
gsets = readLines(con = "set_results/c2.all.v6.0.symbols.gmt")
gsets = strsplit(x=gsets, split = "\t")
names(gsets) = sapply(X = gsets, FUN = function(x) x[1])
gsets = sapply(X = gsets, FUN = function(x) x[3:length(x)])


DEgenes = allresults[(pvalue.ecad_norm.DESeq2 <= 0.05), geneid]
DEgenes = paste("gene:", DEgenes, sep="")


mat <- zMatrix[DEgenes,]
annodf = data.table(genename=rownames(mat))
setkey(annodf, genename)
for (setname in annosets) {
  mgenes = hggenes[gsets[setname], Gene.stable.ID]
  mgenes <- mgenes[!is.na(mgenes)]
  mgenes = paste("gene:", mgenes, sep = "")
  annodf[,eval(setname) := FALSE]
  annodf[mgenes, eval(setname) := TRUE]
}


ann2 = annodf[,-"genename"]
ann2$KEGG_APOPTOSIS = factor(ann2$KEGG_APOPTOSIS)
# ann2$KEGG_APOPTOSIS[ann2$KEGG_APOPTOSIS == FALSE] <- NA
ann2$KEGG_WNT_SIGNALING_PATHWAY = factor(ann2$KEGG_WNT_SIGNALING_PATHWAY)
ann2$KEGG_TGF_BETA_SIGNALING_PATHWAY = factor(ann2$KEGG_TGF_BETA_SIGNALING_PATHWAY)

colnames(ann2)[colnames(ann2) == "KEGG_APOPTOSIS"] = "Apoptosis"
colnames(ann2)[colnames(ann2) == "KEGG_WNT_SIGNALING_PATHWAY"] = "Wnt Signaling"
colnames(ann2)[colnames(ann2) == "KEGG_TGF_BETA_SIGNALING_PATHWAY"] = "TGF-beta Signaling"


hm = annHeatmap(t(mat),
                annotation = ann2,
                dendrogram = list(Col=list(status = "hidden"),Row=list(status="no")),
                labels = list(Col=list(labels=FALSE)),
                breaks = c(seq(from=-3, to=-.5, length.out = 100),
                           seq(from=-.5, to=.5, length.out = 300),
                           seq(from=.5, to=3, length.out = 100)),
                col = colorRampPalette(c("blue","yellow","red"))(499))

pdf("heatmap-pval.05.pdf")
plot(hm)
dev.off()
```



```{r heatmap-FWER.05, include=FALSE, message=FALSE}
annosets = c("KEGG_WNT_SIGNALING_PATHWAY","KEGG_APOPTOSIS","KEGG_TGF_BETA_SIGNALING_PATHWAY")
gsets = readLines(con = "set_results/c2.all.v6.0.symbols.gmt")
gsets = strsplit(x=gsets, split = "\t")
names(gsets) = sapply(X = gsets, FUN = function(x) x[1])
gsets = sapply(X = gsets, FUN = function(x) x[3:length(x)])


fwer = .05/length(unique(allresults[pvalue.ecad_norm.DESeq2 < 1, geneid]))
DEgenes = allresults[ pvalue.ecad_norm.DESeq2 <= fwer, geneid]
DEgenes = paste("gene:", DEgenes, sep="")


mat <- zMatrix[DEgenes,]
annodf = data.table(genename=rownames(mat))
setkey(annodf, genename)
for (setname in annosets) {
  mgenes = hggenes[gsets[setname], Gene.stable.ID]
  mgenes <- mgenes[!is.na(mgenes)]
  mgenes = paste("gene:", mgenes, sep = "")
  annodf[,eval(setname) := FALSE]
  annodf[mgenes, eval(setname) := TRUE]
}


ann2 = annodf[,-"genename"]
ann2$KEGG_APOPTOSIS = factor(ann2$KEGG_APOPTOSIS)
# ann2$KEGG_APOPTOSIS[ann2$KEGG_APOPTOSIS == FALSE] <- NA
ann2$KEGG_WNT_SIGNALING_PATHWAY = factor(ann2$KEGG_WNT_SIGNALING_PATHWAY)
ann2$KEGG_TGF_BETA_SIGNALING_PATHWAY = factor(ann2$KEGG_TGF_BETA_SIGNALING_PATHWAY)

colnames(ann2)[colnames(ann2) == "KEGG_APOPTOSIS"] = "Apoptosis"
colnames(ann2)[colnames(ann2) == "KEGG_WNT_SIGNALING_PATHWAY"] = "Wnt Signaling"
colnames(ann2)[colnames(ann2) == "KEGG_TGF_BETA_SIGNALING_PATHWAY"] = "TGF-beta Signaling"


hm = regHeatmap(mat[,9:1],
                dendrogram = list(Row=list(status = "hidden"),Col=list(status="no")),
                labels = list(Row=list(labels=FALSE)),
#                breaks = c(seq(from=-3, to=-.5, length.out = 100),
#                           seq(from=-.5, to=.5, length.out = 100),
#                           seq(from=.5, to=3, length.out = 100)),
                breaks = c(seq(from=-3,to=3,length.out = 300)),
                #col = colorRampPalette(c("blue","white","red"))(299))
                col = kovesi.diverging_bwr_40_95_c42(299))



pdf("heatmap-fwer.05.pdf")
plot(hm)
dev.off()

```


```{r heatmap-FWER-1, include=FALSE, message=FALSE}
annosets = c("KEGG_WNT_SIGNALING_PATHWAY","KEGG_APOPTOSIS","KEGG_TGF_BETA_SIGNALING_PATHWAY")
gsets = readLines(con = "set_results/c2.all.v6.0.symbols.gmt")
gsets = strsplit(x=gsets, split = "\t")
names(gsets) = sapply(X = gsets, FUN = function(x) x[1])
gsets = sapply(X = gsets, FUN = function(x) x[3:length(x)])


fwer = 1.0/length(unique(allresults[pvalue.ecad_norm.DESeq2 < 1, geneid]))
DEgenes = allresults[pvalue.ecad_norm.DESeq2 <= fwer, geneid]
DEgenes = paste("gene:", DEgenes, sep="")


mat <- zMatrix[DEgenes,]
annodf = data.table(genename=rownames(mat))
setkey(annodf, genename)
for (setname in annosets) {
  mgenes = hggenes[gsets[setname], Gene.stable.ID]
  mgenes <- mgenes[!is.na(mgenes)]
  mgenes = paste("gene:", mgenes, sep = "")
  annodf[,eval(setname) := FALSE]
  annodf[mgenes, eval(setname) := TRUE]
}


ann2 = annodf[,-"genename"]
ann2$KEGG_APOPTOSIS = factor(ann2$KEGG_APOPTOSIS)
# ann2$KEGG_APOPTOSIS[ann2$KEGG_APOPTOSIS == FALSE] <- NA
ann2$KEGG_WNT_SIGNALING_PATHWAY = factor(ann2$KEGG_WNT_SIGNALING_PATHWAY)
ann2$KEGG_TGF_BETA_SIGNALING_PATHWAY = factor(ann2$KEGG_TGF_BETA_SIGNALING_PATHWAY)

colnames(ann2)[colnames(ann2) == "KEGG_APOPTOSIS"] = "Apoptosis"
colnames(ann2)[colnames(ann2) == "KEGG_WNT_SIGNALING_PATHWAY"] = "Wnt Signaling"
colnames(ann2)[colnames(ann2) == "KEGG_TGF_BETA_SIGNALING_PATHWAY"] = "TGF-beta Signaling"


hm = regHeatmap(t(mat),
                dendrogram = list(Col=list(status = "hidden"),Row=list(status="no")),
                labels = list(Col=list(labels=FALSE)),
#                breaks = c(seq(from=-3, to=-.5, length.out = 100),
#                           seq(from=-.5, to=.5, length.out = 100),
#                           seq(from=.5, to=3, length.out = 100)),
                breaks = c(seq(from=-3,to=3,length.out = 300)),
                col = colorRampPalette(c("blue","white","red"))(299))

pdf("heatmap-fwer-1.pdf")
plot(hm)
dev.off()

```


```{r EMT-heatmap-1}
EMTgenes = c("Cdh1", "Twist1", "Snai1", "Fn1", "Zeb1", "Snai2", "Vim", "Chd2", "Cdh3", "Tcf3", "Zeb2", "Foxc2", "Twist2")
EMTgenes <- toupper(EMTgenes)
EMTgenes <- EMTgenes[order(-allresults[hggenes[EMTgenes,Gene.stable.ID],pvalue.ecad_norm.DESeq2])]
setkey(hggenes, HGNC.symbol)
zEMT <- zMatrix[paste0("gene:",hggenes[EMTgenes, Gene.stable.ID]),]
rownames(zEMT) <- EMTgenes
hm = regHeatmap(zEMT,
           dendrogram = list(status = "no"),
                breaks = c(seq(from=-3, to=-.3, length.out = 100), seq(from=-.3, to=.3, length.out = 100),
                           seq(from=.3, to=3, length.out = 100)),
                col = colorRampPalette(c("blue","white","red"))(299))
#pdf("EMTheatmap-1.pdf")
#plot(hm)
#dev.off()
```

```{r EMT-heatmap-2}
dev.off()
dev.new()
EMTgenes = c("Cdh1", "Twist1", "Snai1", "Fn1", "Zeb1", "Snai2", "Vim", "Chd2", "Cdh3", "Tcf3", "Zeb2", "Foxc2", "Twist2")
EMTgenes <- toupper(EMTgenes)
EMTgenes <- EMTgenes[order(-allresults[hggenes[EMTgenes,Gene.stable.ID],pvalue.ecad_norm.DESeq2])]
setkey(hggenes, HGNC.symbol)
zEMT <- zMatrix[paste0("gene:",hggenes[EMTgenes, Gene.stable.ID]),]
rownames(zEMT) <- EMTgenes
hm = regHeatmap(zEMT,
                scale = "none",
           dendrogram = list(status = "no"),
                breaks = c(seq(from=-3, to=-.2, length.out = 100),
                           seq(from=-.2, to=.2, length.out = 300),
                           seq(from=.2, to=3, length.out = 100)),
                col = colorRampPalette(c("blue","white","red"))(499))
pdf("EMTheatmap.pdf")
plot(hm)
dev.off()
```

```{r EMT-heatmap-3}
dev.off()
EMTgenes = c("Cdh1", "Twist1", "Snai1", "Fn1", "Zeb1", "Snai2", "Vim", "Chd2", "Cdh3", "Tcf3", "Zeb2", "Foxc2", "Twist2")
EMTgenes <- toupper(EMTgenes)
EMTgenes <- EMTgenes[order(-allresults[hggenes[EMTgenes,Gene.stable.ID],pvalue.ecad_norm.DESeq2])]
setkey(hggenes, HGNC.symbol)
zEMT <- zMatrix[paste0("gene:",hggenes[EMTgenes, Gene.stable.ID]),]
rownames(zEMT) <- EMTgenes
hm = regHeatmap(zEMT,
           dendrogram = list(status = "no"),
                breaks = c(seq(from=-3, to=-.2, length.out = 100),
                           seq(from=-.2, to=.2, length.out = 800),
                           seq(from=.2, to=3, length.out = 100)),
                col = colorRampPalette(c("blue","white","red"))(999))
#pdf("EMTheatmap-3.pdf")
#plot(hm)
#dev.off()
```

```{r}
krtgenes = c("KRT8",   "KRT9",   "KRT14",   "KRT18",   "CDH1",  "EPCAM", "OCLN",  "TJP3",   "CGN",   "DSP", "SDC1")
setkey(hggenes, HGNC.symbol)

krtgenes <- krtgenes[order(-allresults[hggenes[krtgenes,Gene.stable.ID],pvalue.ecad_norm.DESeq2])]
setkey(hggenes, HGNC.symbol)
zKRT <- zMatrix[paste0("gene:",hggenes[krtgenes, Gene.stable.ID]),]
rownames(zKRT) <- krtgenes
hm = regHeatmap(zKRT,
           dendrogram = list(status = "no"),
           scale = "none",
                breaks = c(seq(from=-3, to=-.2, length.out = 100), 
                           seq(from=-.2, to=.2, length.out = 300),
                           seq(from=.2, to=3, length.out = 100)),
                col = colorRampPalette(c("blue","white","red"))(499))
pdf("KRTheatmap-pval.pdf")
plot(hm)
dev.off()

hm = regHeatmap(zKRT,
           dendrogram = list(Col=list(status = "no")),
           scale = "none",
                breaks = c(seq(from=-3, to=-.2, length.out = 100),
                           seq(from=-.2, to=.2, length.out = 300),
                           seq(from=.2, to=3, length.out = 100)),
                col = colorRampPalette(c("blue","white","red"))(499))
pdf("KRTheatmap-clustered.pdf")
plot(hm)
dev.off()
```


```{r write-certain-genes}
panelEgenes <- c("Casp4","Mcl1","Bmp2","Bcl2l1","Jun","Smad7",
                 "Hspb1","Rela","Rhob","Erbb2","Ereg","Clu",
                 "App","Psen1","Timp3","Anxa1","F2r")
panelEgenes <- toupper(panelEgenes)
wb = openxlsx::createWorkbook(creator = "Yasir Suhail")
openxlsx::addWorksheet(wb, sheetName = "Selected Genes")
openxlsx::writeDataTable(wb = wb, sheet = 1, x = allresults[hggenes[panelEgenes, Gene.stable.ID],])
openxlsx::saveWorkbook(wb, file = "selectedGenes.xlsx", overwrite = T)
```



```{r deleteme}
save.image(file = "deleteme.Rdata")
```

```{r writegenesets}
c2res <- as.data.table(openxlsx::read.xlsx(xlsxFile = "set_results/All KO vs Ecad GSEA.xlsx", sheet="c2 (Curated Gene Sets)"))

hres <- as.data.table(openxlsx::read.xlsx(xlsxFile = "set_results/All KO vs Ecad GSEA.xlsx", sheet="h (Hallmarks)"))

posnames <- c2res[(padj.Cdh1_norm.deseq2 <= .05) | (padj.KO.deseq2 <= .05)][grep(pattern = "KEGG",x = geneset), geneset]

posnames <- c(posnames,
              hres[(padj.Cdh1_norm.deseq2 <= .05) | (padj.KO.deseq2 <= .05), geneset])

gsets_h = readLines(con = "set_results/h.all.v6.0.symbols.gmt")
gsets_h = strsplit(x=gsets_h, split = "\t")
names(gsets_h) = sapply(X = gsets_h, FUN = function(x) x[1])
gsets_h = sapply(X = gsets_h, FUN = function(x) x[3:length(x)])





gsets <- c(gsets, gsets_h)
setkey(hggenes, HGNC.symbol)
setkey(allresults, HGNC.symbol)
for (setn in posnames) {
  humnames <- gsets[[setn]]
  mousenames <- hggenes[humnames, Gene.stable.ID]
  mousenames <- mousenames[! is.na(mousenames)]
  setdata <- allresults[humnames,]
  sumtab <- data.frame(Field=NA,Value=NA)
  sumtab <- rbind(sumtab, list("Number of gene names in the set (Human)",length(unique(humnames))))
  sumtab <- rbind(sumtab, list("Number of mouse ensembl genes in the set",length(unique(mousenames))))
  
  
  sumtab <- rbind(sumtab, list("Num of genes with KO padj <= .05",sum( setdata$FDR.KO <= .05)))
  sumtab <- rbind(sumtab, list("Num of genes with scaled Cdh1 linear model padj <= .05",
                               sum( setdata$FDR.ecad_norm <= .05)))
  
  sumtab <- rbind(sumtab, list("Num of genes with KO padj < 1",sum( setdata$FDR.KO < 1)))
  sumtab <- rbind(sumtab, list("Num of genes with scaled Cdh1 linear model padj < 1",
                               sum( setdata$FDR.ecad_norm < 1)))

  sumtab <- rbind(sumtab, list("Mean z KO",mean( setdata$z.KO)))
  sumtab <- rbind(sumtab, list("Mean z, scaled Cdh1 linear model",mean( setdata$z.ecad_norm)))

  
  
    
  sumtab <- rbind(sumtab, list("Num of genes with l2fc KO < 0",sum( setdata$l2fc.KO.DESeq2 < 0)))
  sumtab <- rbind(sumtab, list("Num of genes with l2fc KO > 0",sum( setdata$l2fc.KO.DESeq2 > 0)))  

  sumtab <- rbind(sumtab, list("Num of genes with l2fc, scaled Cdh1 < 0",
                               sum( setdata$l2fc.ecad_norm.DESeq2 < 0)))
  sumtab <- rbind(sumtab, list("Num of genes with l2fc, scaled Cdh1 > 0",
                               sum( setdata$l2fc.ecad_norm.DESeq2 > 0)))  

  sumtab <- rbind(sumtab, list("Num of genes with l2fc, scaled Cdh1 < 0",
                               sum( setdata$l2fc.ecad_norm.DESeq2 < 0)))
  sumtab <- rbind(sumtab, list("Num of genes with l2fc, scaled Cdh1 > 0",
                               sum( setdata$l2fc.ecad_norm.DESeq2 > 0)))  

  sumtab <- rbind(sumtab, list("Num of genes with l2fc, scaled Cdh1 < 0",
                               sum( setdata$l2fc.ecad_norm.DESeq2 < 0)))
  sumtab <- rbind(sumtab, list("Num of genes with l2fc, scaled Cdh1 > 0",
                               sum( setdata$l2fc.ecad_norm.DESeq2 > 0)))  

  sumtab <- rbind(sumtab, list("Up genes KO (l2fc>0, padj <= .05)=",
                               paste0(setdata[order(-abs(z.KO)),][FDR.KO <= .05][l2fc.KO.DESeq2>0 ]$mgi_symbol, collapse = ", ")))
  sumtab <- rbind(sumtab, list("Up genes KO (l2fc>0, padj < .05), z score=",
                               paste0(sprintf("%.2f",
                                             setdata[order(-abs(z.KO)),][FDR.KO <= .05][l2fc.KO.DESeq2>0 ]$z.KO),
                                     collapse = ", ")))


  sumtab <- rbind(sumtab, list("Down genes KO (l2fc<0, padj <= .05)=",
                               paste0(setdata[order(-abs(z.KO)),][FDR.KO <= .05][l2fc.KO.DESeq2<0 ]$mgi_symbol, collapse = ", ")))
  sumtab <- rbind(sumtab, list("Down genes KO (l2fc<0, padj < .05), z score=",
                               paste0(sprintf("%.2f",
                                             setdata[order(-abs(z.KO)),][FDR.KO <= .05][l2fc.KO.DESeq2<0 ]$z.KO),
                                     collapse = ", ")))

  
  
  sumtab <- rbind(sumtab, list("Up genes scaled Cdh1 (l2fc>0, padj <= .05)=",
                               paste0(setdata[order(-abs(z.ecad_norm)),][FDR.ecad_norm <= .05][l2fc.ecad_norm.DESeq2>0 ]$mgi_symbol, collapse = ", ")))
  sumtab <- rbind(sumtab, list("Up genes scaled Cdh1 (l2fc>0, padj < .05), z score=",
                               paste0(sprintf("%.2f",
                                             setdata[order(-abs(z.ecad_norm)),][FDR.ecad_norm <= .05][l2fc.ecad_norm.DESeq2>0 ]$z.KO),
                                     collapse = ", ")))
  

  sumtab <- rbind(sumtab, list("Down genes scaled Cdh1 (l2fc<0, padj <= .05)=",
                               paste0(setdata[order(-abs(z.ecad_norm)),][FDR.ecad_norm <= .05][l2fc.ecad_norm.DESeq2<0 ]$mgi_symbol, collapse = ", ")))
  sumtab <- rbind(sumtab, list("Down genes scaled Cdh1 (l2fc<0, padj < .05), z score=",
                               paste0(sprintf("%.2f",
                                             setdata[order(-abs(z.ecad_norm)),][FDR.ecad_norm <= .05][l2fc.ecad_norm.DESeq2<0 ]$z.KO),
                                     collapse = ", ")))
  

  wb = openxlsx::createWorkbook(creator = "Yasir Suhail")
  openxlsx::addWorksheet(wb, sheetName = "summary")
  openxlsx::addWorksheet(wb, sheetName = "data")
  openxlsx::writeDataTable(wb=wb,sheet="summary",x=sumtab)
  openxlsx::writeDataTable(wb=wb,sheet="data",x=setdata[order(-abs(z.KO)),])
  openxlsx::saveWorkbook(wb, file = paste0("set_results/gene_sets/",setn,".xlsx"), overwrite = T)
}

```




```{r write-allresults-with-HGNC}
rr = allresults
rr = rr[, HGNC.symbols := paste(unique(HGNC.symbol), collapse = ","), by=geneid]
rr = rr[, mgi_symbols := paste(unique(mgi_symbol), collapse = ","), by=geneid]
rr = unique(rr[,-c("HGNC.symbol","mgi_symbol")])


openxlsx::write.xlsx(rr, file = "allresults_withHGNC.xlsx", 
                     sheetName = "DE Gene-level results", overwrite=TRUE,
                     asTable = T)

apo_dt <- rr[grep(pattern = "apoptosis", x=description),]

openxlsx::write.xlsx(apo_dt, file = "apoptosis_in_description.xlsx", 
                     sheetName = "DE Gene-level results", overwrite=TRUE,
                     asTable = T)


rm(rr)

```



```{r}
library("pathview")


ensembl2entrez <- getBM(attributes = c("ensembl_gene_id","entrezgene"), mart = ensembl)
ensembl2entrez <- as.data.frame(ensembl2entrez)
allresults <- merge(allresults, ensembl2entrez, by.x="geneid", by.y="ensembl_gene_id", all.x=T)

expr = allresults$l2fc.ecad_norm.DESeq2
names(expr) <- allresults$entrezgene
expr <- expr[! is.na(names(expr))]


plotKEGG <- function(expr, pathway="04210", suffix="linear-l2fc", ...) {
  pv.out <- pathview(gene.data = expr, pathway.id = pathway, 
                   species = "mmu", out.suffix = suffix, kegg.native = T, same.layer = F, ...)
}

plotKEGG(expr, limit = list(gene = .6, cpd=1))
plotKEGG(expr, pathway = "04390", limit = list(gene = .6, cpd=1))
```



```{r write-results-with-apoptosis}

hsets = readLines(con = "set_results/h.all.v6.0.symbols.gmt")
hsets = strsplit(x=hsets, split = "\t")
names(hsets) = sapply(X = hsets, FUN = function(x) x[1])
hsets = sapply(X = hsets, FUN = function(x) x[3:length(x)])

h_apo_genes = hsets[["HALLMARK_APOPTOSIS"]]
g_apo_genes = gsets[["KEGG_APOPTOSIS"]]




grep(pattern = "APOPTOSIS", x = names(hsets), value = T)

allresults$KEGG_APOPTOSIS = 0
allresults$HALLMARK_APOPTOSIS = 0
allresults[HGNC.symbol %in% g_apo_genes, KEGG_APOPTOSIS := 1]
allresults[HGNC.symbol %in% h_apo_genes, HALLMARK_APOPTOSIS := 1]

openxlsx::write.xlsx(allresults, file = "allresults_with_apoptosis_columns.xlsx", 
                     sheetName = "DE Gene-level results", overwrite=TRUE,
                     asTable = T)


```


```{r}
library(pals)
classical_cadherins <- paste0("Cdh", c(1:12,15,18:20, 22,24))
desmo_cadherins <- c("Dsg1","Dsg2","Dsg3","Dsg4", "Dsc1","Dsc2","Dsc3")
cadherins <- data.frame(hgnc.symbol = c(classical_cadherins, desmo_cadherins),
                        family=c(rep("classical", length(classical_cadherins)), 
                                 rep("desmosomal", length(desmo_cadherins))))


#cadherins <- data.frame(HGNC.symbol=c("CDH1","CDH2","CDH12","CDH3", "DSG1","DSG2","DSG3","DSG4", "DSC1","DSC2","DSC3"),
#                        family=c(rep("classical", 4), rep("desmosomal", 7)))

get_heatmap_from_genes <- function(hsymbol_tab, result_object) {
  setkey(hggenes, "HGNC.symbol")
  hsymbol_tab$HGNC.symbol <- toupper(hsymbol_tab$hgnc.symbol)
  hsymbol_tab <- hsymbol_tab[hsymbol_tab$HGNC.symbol %in% hggenes$HGNC.symbol, ]
  gene_tab <- merge(hsymbol_tab, hggenes, by="HGNC.symbol", all.x=T)
  
  gene_tab$Gene.stable.ID_gene <- paste0("gene:",gene_tab$Gene.stable.ID)
  gene_tab$pvalue <- result_object[gene_tab$Gene.stable.ID_gene,]$pvalue
  gene_tab <- gene_tab[rev(order(gene_tab$pvalue)),]
  gene_mat <- zMatrix[gene_tab$Gene.stable.ID_gene,c(9:1)]
  rownames(gene_mat) <- gene_tab$hgnc.symbol
  gene_mat <- gene_mat[! is.na(rowSums(gene_mat)),]
  
  # a = regHeatmap(gene_mat, dendrogram = list(Col=list(status = "no"), Row=list(status="no")), 
  #                scale = "none",
  #                breaks = c(seq(from=-3, to=-.2, length.out = 100),
  #                           seq(from=-.2, to=.2, length.out = 300),
  #                           seq(from=.2, to=3, length.out = 100)),
  #                col = colorRampPalette(c("blue","white","red"))(499)
  # )
  a = regHeatmap(gene_mat, dendrogram = list(Col=list(status = "no"), Row=list(status="no")),
                 labels = list(Row=list(side=2)),
                 scale = "none",
                 breaks = seq(from=-3, to=3, length.out = 500),
                 #col = colorRampPalette(c("blue","white","red"))(499),
                 col = kovesi.diverging_bwr_40_95_c42(499),
                 legend = 1
  )

  return(a)
  
}
 
myplot.annHeatmap <- function (x, widths, heights, mmar, omar, ...) 
{
    opar = par("oma", "mar", "xaxs", "yaxs")
    on.exit(par(opar))
    doRClusLab = !is.null(x$cluster$Row$label)
    doCClusLab = !is.null(x$cluster$Col$label)
    if (missing(omar))
      omar = rep(0, 4)
    if (doRClusLab) 
        omar[1] = 2
    if (doCClusLab) 
        omar[4] = 2
    par(oma = omar)
    if (!missing(widths)) 
        x$layout$width = widths
    if (!missing(heights)) 
        x$layout$height = heights
    with(x$layout, layout(plot, width, height, respect = TRUE))
    nc = ncol(x$data$x2)
    nr = nrow(x$data$x2)
    doRlab = !is.null(x$labels$Row$labels)
    doClab = !is.null(x$labels$Col$labels)
    if (missing(mmar))
      mmar = c(1, 0, 0, 2)
    if (doRlab) 
        mmar[x$labels$Row$side] = 1.5*x$labels$Row$nrow
    if (doClab) 
        mmar[x$labels$Col$side] = x$labels$Col$nrow
    with(x$data, {
        par(mar = mmar)
        image(1:nc, 1:nr, t(x2), axes = FALSE, xlim = c(0.5, 
            nc + 0.5), ylim = c(0.5, nr + 0.5), xlab = "", ylab = "", 
            col = col, breaks = breaks, ...)
    })
    with(x$labels, {
        if (doRlab) 
            axis(Row$side, 1:nr, las = 2, line = -0.5, tick = 0, 
                labels = Row$labels, cex.axis = Row$cex)
        if (doClab) 
            axis(Col$side, 1:nc, las = 2, line = -0.5, tick = 0, 
                labels = Col$labels, cex.axis = Col$cex)
    })
    with(x$dendrogram$Col, if (status == "yes") {
        par(mar = c(0, mmar[2], 3, mmar[4]))
        cutplot.dendrogram(dendro, h = x$cluster$Col$cuth, cluscol = x$cluster$Col$col, 
            horiz = FALSE, axes = FALSE, xaxs = "i", leaflab = "none", 
            lwd = x$dendrogram$Col$lwd)
    })
    with(x$dendrogram$Row, if (status == "yes") {
        par(mar = c(mmar[1], 3, mmar[3], 0))
        cutplot.dendrogram(dendro, h = x$cluster$Row$cuth, cluscol = x$cluster$Row$col, 
            horiz = TRUE, axes = FALSE, yaxs = "i", leaflab = "none", 
            lwd = x$dendrogram$Row$lwd)
    })
    if (!is.null(x$annotation$Col$data)) {
        par(mar = c(1, mmar[2], 0, mmar[4]), xaxs = "i", yaxs = "i")
        picketPlot(x$annotation$Col$data[x$data$colInd, , drop = FALSE], 
            grp = x$cluster$Col$grp, grplabel = x$cluster$Col$label, 
            grpcol = x$cluster$Col$col, control = x$annotation$Col$control, 
            asIs = TRUE)
    }
    if (!is.null(x$annotation$Row$data)) {
        par(mar = c(mmar[1], 0, mmar[3], 1), xaxs = "i", yaxs = "i")
        picketPlot(x$annotation$Row$data[x$data$rowInd, , drop = FALSE], 
            grp = x$cluster$Row$grp, grplabel = x$cluster$Row$label, 
            grpcol = x$cluster$Row$col, control = x$annotation$Row$control, 
            asIs = TRUE, horizontal = FALSE)
    }
    if (x$legend) {
        if (x$layout$legend.side %in% c(1, 3)) {
            par(mar = c(2, mmar[2] + 2, 2, mmar[4] + 2))
        }
        else {
            par(mar = c(mmar[1] + 2, 2, mmar[3] + 2, 2))
        }
        doLegend(x$data$breaks, col = x$data$col, x$layout$legend.side)
    }
    invisible(x)
}
 

pdf("Cadherins-ecad-norm.pdf")
par(mar=c(4,4,4,4)*4)
myplot.annHeatmap(get_heatmap_from_genes(hsymbol_tab = cadherins, result_object = results.norm_ecad.DESeq2), omar =rep(1,4))
dev.off()


pdf("Classical_cadherins-ecad-norm.pdf")
par(mar=c(4,4,4,4)*4)
cd1 <- cadherins[cadherins$family=="classical",]
myplot.annHeatmap(get_heatmap_from_genes(hsymbol_tab = cd1, result_object = results.norm_ecad.DESeq2), omar =rep(1,4))
dev.off()  


pdf("desmosomal_cadherins-ecad-norm.pdf")
par(mar=c(4,4,4,4)*4)
cd2 <- cadherins[cadherins$family=="desmosomal",]
myplot.annHeatmap(get_heatmap_from_genes(hsymbol_tab = cd2, result_object = results.norm_ecad.DESeq2), omar =rep(1,4))
dev.off()

pdf("EMTgenes-ecad-norm.pdf")
par(mar=c(4,4,4,4)*4)
EMTgenes = c("Cdh1", "Twist1", "Snai1", "Fn1", "Zeb1", "Snai2", "Vim", "Chd2", "Cdh3", "Tcf3", "Zeb2", "Foxc2", "Twist2")
EMTtab <- data.frame(hgnc.symbol = EMTgenes, family="EMT")
myplot.annHeatmap(get_heatmap_from_genes(hsymbol_tab = EMTtab, result_object = results.norm_ecad.DESeq2), omar =rep(1,4))
dev.off()


pdf("Epithelial-ecad-norm.pdf")
par(mar=c(4,4,4,4)*4)
epigenes = c("Krt8",   "Krt9",   "Krt14",   "Krt18",   "Cdh1",  "EpCAM", "Ocln",  "Tjp3",   "Cgn",   "Dsp", "Sdc1")
epitab <- data.frame(hgnc.symbol = epigenes, family="Epithelial")
myplot.annHeatmap(get_heatmap_from_genes(hsymbol_tab = epitab, result_object = results.norm_ecad.DESeq2), omar =rep(1,4))
dev.off()

pdf("EMT-epi-ecad-norm.pdf")
par(mar=c(4,4,4,4)*4)
EMT_epigenes = c("Cdh1", "Twist1", "Snai1", "Fn1", "Zeb1", "Snai2", "Vim", "Chd2", "Cdh3", "Tcf3", "Zeb2", "Foxc2", "Twist2", 
                 "Krt8",   "Krt9",   "Krt14",   "Krt18",  "EpCAM", "Ocln",  "Tjp3",   "Cgn",   "Dsp", "Sdc1")
EMT_epitab <- data.frame(hgnc.symbol = EMT_epigenes, family="EMT")
myplot.annHeatmap(get_heatmap_from_genes(hsymbol_tab = EMT_epitab, result_object = results.norm_ecad.DESeq2), omar =rep(1,4))
dev.off()




setkey(hggenes, "HGNC.symbol")
cadherins <- merge(cadherins, hggenes, by="HGNC.symbol", all.x=T)

cadherins$Gene.stable.ID_gene <- paste0("gene:",cadherins$Gene.stable.ID)

cad_mat <- zMatrix[cadherins$Gene.stable.ID_gene,]
rownames(cad_mat) <- cadherins$HGNC.symbol
cad_mat <- cad_mat[! is.na(rowSums(cad_mat)),]
jpeg("Cadherins.jpg")
a = regHeatmap(cad_mat, dendrogram = list(Col=list(status = "no"), Row=list(status="yes")), 
               scale = "none",
               breaks = c(seq(from=-3, to=-.2, length.out = 100),
                          seq(from=-.2, to=.2, length.out = 300),
                          seq(from=.2, to=3, length.out = 100)),
               col = colorRampPalette(c("blue","white","red"))(499)
               )
plot(a)
dev.off()



cad_p <- allresults[allresults$geneid %in% cadherins$Gene.stable.ID,c("HGNC.symbol","pvalue.KO.DESeq2", "FDR.KO", "l2fc.KO.DESeq2")]
cad_p <- cad_p[order(HGNC.symbol),]
cad_p

```


```{r}
tgfgenes <- gsets$KEGG_TGF_BETA_SIGNALING_PATHWAY
tgf.df <- hggenes[tgfgenes,]
tgf.df <- tgf.df[! is.na(tgf.df$Gene.stable.ID),]

tgf.df$Gene.stable.ID_gene <- paste0("gene:",tgf.df$Gene.stable.ID)

tgftab <- data.frame(hgnc.symbols = tgfgenes)
pdf("tgf.pdf")
hm <- get_heatmap_from_genes(hsymbol_tab = tgftab, result_object = results.norm_ecad.DESeq2)
hm$cluster$Row$cex = 0.5
myplot.annHeatmap(hm, omar =rep(1,4))
dev.off()

tgf_mat <- zMatrix[tgf.df$Gene.stable.ID_gene,]
rownames(tgf_mat) <- tgf.df$HGNC.symbol
tgf_mat <- tgf_mat[! is.na(rowSums(tgf_mat)),]
jpeg("tgfb.jpg")
a = regHeatmap(tgf_mat, dendrogram = list(Col=list(status = "no"), Row=list(status="yes")), 
               scale = "none",
               breaks = c(seq(from=-3, to=-.2, length.out = 100),
                          seq(from=-.2, to=.2, length.out = 300),
                          seq(from=.2, to=3, length.out = 100)),
               col = colorRampPalette(c("blue","white","red"))(499)
               )
plot(a)
dev.off()


```


```{r cadherin_lm}

cad_results <- merge(cadherins, allresults, by="HGNC.symbol")
cad_results$z.KO[cad_results$z.KO ==0] <- NA
cad_results[cad_results$HGNC.symbol=="CDH1",]$z.KO <- NA
cad_results[cad_results$HGNC.symbol=="CDH1",]$l2fc.KO.DESeq2 <- NA
summary(lm(cad_results$l2fc.KO.DESeq2 ~ cad_results$family))

```

## Notes and diagnostics:



### Comparison of the methods


```{r, warning=FALSE, fig.height= 11, fig.width=11, fig.cap="Pairwise comparison of p-values"}
library(GGally)
ggpairs(allresults[, grep(pattern = "pvalue", x = colnames(allresults), value = TRUE), with=FALSE]) 

```

```{r, warning=FALSE, fig.height= 11, fig.width=11, fig.cap="pairwise comparison of -log10 pvalues"}
library(GGally)
ggpairs(-log10(allresults[, grep(pattern = "pvalue", x = colnames(allresults), value = TRUE), with=FALSE]))
```


```{r, warning=FALSE, fig.height= 11, fig.width=11, fig.cap="Pairwise comparison of log2 fold change"}
library(GGally)
ggpairs(allresults[, grep(pattern = "l2fc", x = colnames(allresults), value = TRUE), with=FALSE])
```


```{r}
save.image("allData.RData")
```



# References
